shared/docs/PROTOCOL_STATE_HASH.md

```diff
@@ -0,0 +1,173 @@
+
+    %% Handle New Message
+    B-->>C: HandleMessage(msg)
+
+    critical Decide
+        %% Commit Context
+        C->>+U: Context.Commit(quorumCert)
+        U->>+P: Context.Commit(quorumCert)
+        P->>P: Internal Implementation
+        Note over P: Store Block
+        P->>-U: err_code
+        U->>C: err_code
+        deactivate U
+
+        %% Release Context
+        C->>+U: Context.Release()
+        U->>+P: Context.Release()
+        P->>-U: err_code
+        U->>-C: err_code
+    end
+
+```
+
+## Block Application
+
+When applying the block during the `NEWROUND` message shown above, the majority of the flow is similar between the _leader_ and the _replica_ with one of the major differences being a call to the `Utility` module as seen below.
+
+- `ApplyBlock` - Uses the existing set of transactions to validate & propose
+- `CreateAndApplyProposalBlock` - Reaps the mempool for a new set of transaction to validate and propose
+
+```mermaid
+graph TD
+    B[Should I prepare a new block?] --> |Wait for 2/3+ NEWROUND messages| C
+
+    C[Am I the leader?] --> |Yes| D
+    C[Am I the leader?] --> |No| Z
+
+    D[Did I get any prepareQCs?] --> |Find highest valid PrepareQC| E
+    D[Did I get any prepareQCs?] --> |No| Z
+
+    E[Am I ahead of highPrepareQC?] --> |Yes| G
+    E[Am I ahead of highPrepareQC?] --> |No| Z
+
+    G[CreateAndApplyProposalBlock]
+    Z[ApplyBlock]
+```
+
+As either the _leader_ or _replica_, the following steps are followed to apply the proposal transactions in the block.
+
+1.  Retrieve the `PersistenceContext` from the `UtilityContext`
+2.  Update the `PersistenceContext` with the proposed block
+3.  Call either `ApplyBlock` or `CreateAndApplyProposalBlock` based on the flow above
+
+```mermaid
+sequenceDiagram
+    title Steps 1-3
+    participant C as Consensus
+    participant U as Utility
+    participant P as Persistence
+
+        %% Retrieve the persistence context
+        C->>+U: GetPersistenceContext()
+        U->>-C: PersistenceContext
+
+        %% Update the proposal in the persistence context
+        C->>+P: SetProposalBlock
+        P->>-C: err_code
+
+        %% Apply the block to the local proposal state
+        C->>+U: ApplyBlock / CreateAndApplyProposalBlock
+        U->>-C: err_code
+```
+
+4. Loop over all transactions proposed
+5. Check if the transaction has already been applied to the local state
+6. Perform the CRUD operation(s) corresponding to each transaction
+7. The persistence module's internal implementation for [Updating a State hash](../../persistence/docs/PROTOCOL_STATE_HASH.md) must be triggered
+8. Validate that the local state hash computed is the same as that proposed
+
+```mermaid
+sequenceDiagram
+    title Steps 4-8
+    participant C as Consensus
+    participant U as Utility
+    participant P as Persistence
+
+    loop for each tx in txs
+        U->>+P: TransactionExists(txHash)
+        P->>-U: true | false
+        opt if tx is not indexed
```

Comment:
I'm not sure the `opt` is strong enough to show `loop termination` if the `TransactionExists()`

Because ofc if the `TransactionExists()` then the `Block is Invalid` and we can stop `ApplyBlock`

Reply:
I decided to simply change the return value to `false (does not exist)` and remove `opt` altogether.

It only shows the happy path but much clearer IMO.